url56b1dc76d3_wufunxiang_Libs_tgz-pJ8-org_msgpack_util_TemplatePrecompilerJ8/src/org/msgpack/packer/MessagePackPacker.java

```diff
@@ -0,0 +1,334 @@
+//
+// MessagePack for Java
+//
+// Copyright (C) 2009 - 2013 FURUHASHI Sadayuki
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+//
+package org.msgpack.packer;
+
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.math.BigInteger;
+import org.msgpack.io.Output;
+import org.msgpack.io.StreamOutput;
+import org.msgpack.MessagePack;
+import org.msgpack.MessageTypeException;
+
+public class MessagePackPacker extends AbstractPacker {
+    protected final Output out;
+
+    private PackerStack stack = new PackerStack();
+
+    public MessagePackPacker(MessagePack msgpack, OutputStream stream) {
+        this(msgpack, new StreamOutput(stream));
+    }
+
+    protected MessagePackPacker(MessagePack msgpack, Output out) {
+        super(msgpack);
+        this.out = out;
+    }
+
+    
+    protected void writeByte(byte d) throws IOException {
+        if (d < -(1 << 5)) {
+            out.writeByteAndByte((byte) 0xd0, d);
+        } else {
+            out.writeByte(d);
+        }
+        stack.reduceCount();
+    }
+
+    
+    protected void writeShort(short d) throws IOException {
+        if (d < -(1 << 5)) {
+            if (d < -(1 << 7)) {
+                // signed 16
+                out.writeByteAndShort((byte) 0xd1, d);
+            } else {
+                // signed 8
+                out.writeByteAndByte((byte) 0xd0, (byte) d);
+            }
+        } else if (d < (1 << 7)) {
+            // fixnum
+            out.writeByte((byte) d);
+        } else {
+            if (d < (1 << 8)) {
+                // unsigned 8
+                out.writeByteAndByte((byte) 0xcc, (byte) d);
+            } else {
+                // unsigned 16
+                out.writeByteAndShort((byte) 0xcd, d);
+            }
+        }
+        stack.reduceCount();
+    }
+
+    
+    protected void writeInt(int d) throws IOException {
+        if (d < -(1 << 5)) {
+            if (d < -(1 << 15)) {
+                // signed 32
+                out.writeByteAndInt((byte) 0xd2, d);
+            } else if (d < -(1 << 7)) {
+                // signed 16
+                out.writeByteAndShort((byte) 0xd1, (short) d);
+            } else {
+                // signed 8
+                out.writeByteAndByte((byte) 0xd0, (byte) d);
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 82:91, 95:106.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.